// @ui5/webcomponents-react/wrappers/enums/IndicationColor.js
/**
 * Colors to highlight certain UI elements.
 *
 * In contrast to the `ValueState`, the semantic meaning must be defined by the application.
 */
export let IndicationColor = /*#__PURE__*/function (IndicationColor) {
  /**
   * Dark Red
   */
  IndicationColor["Indication01"] = "Indication01";
  /**
   * Red
   */
  IndicationColor["Indication02"] = "Indication02";
  /**
   * Yellow
   */
  IndicationColor["Indication03"] = "Indication03";
  /**
   * Green
   */
  IndicationColor["Indication04"] = "Indication04";
  /**
   * Blue
   */
  IndicationColor["Indication05"] = "Indication05";
  /**
   * Teal
   */
  IndicationColor["Indication06"] = "Indication06";
  /**
   * Purple
   */
  IndicationColor["Indication07"] = "Indication07";
  /**
   * Pink
   */
  IndicationColor["Indication08"] = "Indication08";
  /**
   * Black/White
   */
  IndicationColor["Indication09"] = "Indication09";
  /**
   * Grey
   */
  IndicationColor["Indication10"] = "Indication10";
  /**
   * Dark Red (second color set)
   */
  IndicationColor["Indication11"] = "Indication11";
  /**
   * Red (second color set)
   */
  IndicationColor["Indication12"] = "Indication12";
  /**
   * Yellow (second color set)
   */
  IndicationColor["Indication13"] = "Indication13";
  /**
   * Green (second color set)
   */
  IndicationColor["Indication14"] = "Indication14";
  /**
   * Blue (second color set)
   */
  IndicationColor["Indication15"] = "Indication15";
  /**
   * Teal (second color set)
   */
  IndicationColor["Indication16"] = "Indication16";
  /**
   * Purple (second color set)
   */
  IndicationColor["Indication17"] = "Indication17";
  /**
   * Pink (second color set)
   */
  IndicationColor["Indication18"] = "Indication18";
  /**
   * Black/White (second color set)
   */
  IndicationColor["Indication19"] = "Indication19";
  /**
   * Grey (second color set)
   */
  IndicationColor["Indication20"] = "Indication20";
  return IndicationColor;
}({});
