// @ui5/webcomponents-react/dist/enums/FlexBoxJustifyContent.js
export let FlexBoxJustifyContent = /*#__PURE__*/function (FlexBoxJustifyContent) {
  /**
   * Items are aligned toward the start of the flex-direction.<br />
   * <b>Note:</b> Corresponds to `flex-start`.
   */
  FlexBoxJustifyContent["Start"] = "Start";
  /**
   * Items are centered.<br />
   * <b>Note:</b> Corresponds to `center`.
   */
  FlexBoxJustifyContent["Center"] = "Center";
  /**
   * Items are aligned toward the end of the flex-direction.<br />
   * <b>Note:</b> Corresponds to `flex-end`.
   */
  FlexBoxJustifyContent["End"] = "End";
  /**
   * Items are evenly distributed on the main-axis with equal space around them.<br />
   * <b>Note:</b> Corresponds to `space-around`.
   */
  FlexBoxJustifyContent["SpaceAround"] = "SpaceAround";
  /**
   * Items are evenly distributed on the main-axis; first item is on the start line, last item on the end line.<br />
   * <b>Note:</b> Corresponds to `space-between`.
   */
  FlexBoxJustifyContent["SpaceBetween"] = "SpaceBetween";
  return FlexBoxJustifyContent;
}({});
